remote_config/src/config_update_listener_registration.cc

```diff
@@ -1,34 +1,131 @@
-// Copyright 2023 Google LLC
-//
-// Licensed under the Apache License, Version 2.0 (the "License");
-// you may not use this file except in compliance with the License.
-// You may obtain a copy of the License at
-//
-//      http://www.apache.org/licenses/LICENSE-2.0
-//
-// Unless required by applicable law or agreed to in writing, software
-// distributed under the License is distributed on an "AS IS" BASIS,
-// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-// See the License for the specific language governing permissions and
-// limitations under the License.
+/*
+ * Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
 
 #include "remote_config/src/include/firebase/remote_config/config_update_listener_registration.h"
 
+#include <utility>
+
+#include "app/src/include/firebase/internal/platform.h"
+#include "remote_config/src/cleanup.h"
+#include "remote_config/src/config_update_listener_registration_internal.h"
+
+// The Platform-specific implementation of RemoteConfigInternal is needed
+// so that CleanupFn can reference the cleanup_notifier
+#if FIREBASE_PLATFORM_ANDROID
+#include "remote_config/src/android/remote_config_android.h"
+#elif FIREBASE_PLATFORM_IOS || FIREBASE_PLATFORM_TVOS
+#include "remote_config/src/ios/remote_config_ios.h"
+#else
+#include "remote_config/src/desktop/remote_config_desktop.h"
+#endif  // FIREBASE_PLATFORM_ANDROID, FIREBASE_PLATFORM_IOS,
+        // FIREBASE_PLATFORM_TVOS
+
 namespace firebase {
 namespace remote_config {
+// ConfigUpdateListenerRegistration specific fact:
+//   ConfigUpdateListenerRegistration does NOT own the
+//   ConfigUpdateListenerRegistrationInternal object, which is different from
+//   other wrapper types. RemoteConfigInternal owns all
+//   ConfigUpdateListenerRegistrationInternal objects instead. So
+//   RemoteConfigInternal can remove all listeners upon destruction.
+
+using CleanupFnConfigUpdateListenerRegistration =
+    CleanupFn<ConfigUpdateListenerRegistration, internal::RemoteConfigInternal>;
 
 ConfigUpdateListenerRegistration::ConfigUpdateListenerRegistration(
-    std::function<void()> listener_removal_function)
-    : listener_removal_function_(listener_removal_function) {}
+    const ConfigUpdateListenerRegistration& registration)
+    : remote_config_(registration.remote_config_) {
+  internal_ = registration.internal_;
```

Comment:
this can also be internal_(registration.internal_) after :

Reply:
Done.